pkg/controller/volume/expand/expand_controller.go

```diff
@@ -123,107 +113,156 @@ func NewExpandController(
+			if newSize.Cmp(oldSize) > 0 {
+				expc.enqueuePVC(new)
+			}
+		},
+		DeleteFunc: expc.enqueuePVC,
 	})
 
-	expc.syncResize = NewSyncVolumeResize(syncLoopPeriod, expc.opExecutor, expc.resizeMap, kubeClient)
-	expc.pvcPopulator = NewPVCPopulator(
-		populatorLoopPeriod,
-		expc.resizeMap,
-		expc.pvcLister,
-		expc.pvLister,
-		&expc.volumePluginMgr,
-		kubeClient)
 	return expc, nil
 }
 
-func (expc *expandController) Run(stopCh <-chan struct{}) {
-	defer runtime.HandleCrash()
-	klog.Infof("Starting expand controller")
-	defer klog.Infof("Shutting down expand controller")
-
-	if !controller.WaitForCacheSync("expand", stopCh, expc.pvcsSynced, expc.pvSynced) {
+func (expc *expandController) enqueuePVC(obj interface{}) {
+	pvc, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
 		return
 	}
 
-	// Run volume sync work goroutine
-	go expc.syncResize.Run(stopCh)
-	// Start the pvc populator loop
-	go expc.pvcPopulator.Run(stopCh)
-	<-stopCh
-}
+	size := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	statusSize := pvc.Status.Capacity[v1.ResourceStorage]
 
-func (expc *expandController) deletePVC(obj interface{}) {
-	pvc, ok := obj.(*v1.PersistentVolumeClaim)
-	if !ok {
-		tombstone, ok := obj.(kcache.DeletedFinalStateUnknown)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("couldn't get object from tombstone %+v", obj))
+	if pvc.Status.Phase == v1.ClaimBound && size.Cmp(statusSize) > 0 {
+		key, err := kcache.DeletionHandlingMetaNamespaceKeyFunc(pvc)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pvc, err))
 			return
 		}
-		pvc, ok = tombstone.Obj.(*v1.PersistentVolumeClaim)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("tombstone contained object that is not a pvc %#v", obj))
-			return
+		expc.queue.Add(key)
+	}
+}
+
+func (expc *expandController) processNextWorkItem() bool {
+	key, shutdown := expc.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer expc.queue.Done(key)
+
+	err := expc.syncHandler(key.(string))
+	if err == nil {
+		expc.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	expc.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (expc *expandController) syncHandler(key string) error {
+	namespace, name, err := kcache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	pvc, err := expc.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		klog.V(5).Infof("Error getting PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+
+	pv, err := getPersistentVolume(pvc, expc.pvLister)
+	if err != nil {
+		klog.V(5).Infof("Error getting Persistent Volume for PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+	if pv.Spec.ClaimRef == nil || pvc.Namespace != pv.Spec.ClaimRef.Namespace || pvc.Name != pv.Spec.ClaimRef.Name {
```

Comment:
What about UID?   When does this happen?  We already check at the beginning that PVC is bound right?

Reply:
You're right, checking `pv.spec.claimRef.UID` too would be more robust. I can't think of a case where it would actually help to prevent resizing of a wrong PV, but that does not mean it does not exists. It can race with PV controller with very unexpected ways.

Reply:
okay - are you thinking adding `UID` check in addition to `pvc.Name != pv.Spec.ClaimRef.Name` ? I am going to replace the name check with UID check. I think that should be sufficient. 

Also yeah - we are checking that PVC is bound inside `enquePVC` function.

Reply:
fixed